pkg/watermark/fetch/edge_fetcher.go

```diff
@@ -53,9 +53,40 @@ func NewEdgeFetcher(ctx context.Context, bufferName string, storeWatcher store.W
 	}
 }
 
-// GetHeadWatermark returns the watermark using the HeadOffset (the latest offset among all processors). This
-// can be used in showing the watermark progression for a vertex when not consuming the messages
-// directly (eg. UX, tests)
+// GetWatermark gets the smallest timestamp for the given offset
+func (e *edgeFetcher) GetWatermark(inputOffset isb.Offset) wmb.Watermark {
+	var offset, err = inputOffset.Sequence()
+	if err != nil {
+		e.log.Errorw("Unable to get offset from isb.Offset.Sequence()", zap.Error(err))
+		return wmb.Watermark(time.Unix(-1, 0))
+	}
+	var debugString strings.Builder
+	var epoch int64 = math.MaxInt64
+	var allProcessors = e.processorManager.GetAllProcessors()
+	for _, p := range allProcessors {
+		debugString.WriteString(fmt.Sprintf("[Processor: %v] \n", p))
+		var t = p.offsetTimeline.GetEventTime(inputOffset)
+		if t == -1 { // watermark cannot be computed, perhaps a new processing unit was added or offset fell off the timeline
+			epoch = t
+		} else if t < epoch {
+			epoch = t
+		}
+		if p.IsDeleted() && (offset > p.offsetTimeline.GetHeadOffset()) {
```

Comment:
Should this go before line 68 with `continue`, and then line 70 could be followed by a `break`?